src/coreclr/jit/optimizer.cpp

```diff
@@ -4504,6 +4499,16 @@ bool Compiler::optCompactLoop(FlowGraphNaturalLoop* loop)
             continue;
         }
 
+        // If this is a CALLFINALLYRET that is not in the loop, but the
+        // CALLFINALLY was, then we have to leave it in place. For compaction
+        // purposes this doesn't really make any difference, since no codegen
+        // is associated with the CALLFINALLYRET anyway.
+        if (cur->isBBCallFinallyPairTail())
+        {
+            cur = cur->Next();
+            continue;
+        }
```

Comment:
This is an interesting case that can occur since we do not return any EH succeessors for `CALLFINALLYRET`; if the `CALLFINALLY/CALLFINALLYRET` is inside a try whose handler loops back around, then the `CALLFINALLY` may be part of the loop while the `CALLFINALLYRET` isn't. [Here's an example case](https://edotor.net/?engine=dot?engine=dot?engine=dot#digraph%20FlowGraph%20%7B%0A%20%20%20%20graph%20%5Blabel%20%3D%20%22LocalPopCore%5Cnafter%5CnFind%20loops%22%5D%3B%0A%20%20%20%20node%20%5Bshape%20%3D%20%22Box%22%5D%3B%0A%20%20%20%20BB01%20%5Blabel%20%3D%20%22BB01%5Cn%5Cn%20%20%201.00%22%2C%20shape%20%3D%20%22house%22%5D%3B%0A%20%20%20%20BB02%20%5Blabel%20%3D%20%22BB02%5Cn%5BIND%5D%20%3C%20V01%5Cn%5Cn%20%20%201.00%22%5D%3B%0A%20%20%20%20BB03%20%5Blabel%20%3D%20%22BB03%5Cn%5Cn%20%20%200.00%22%2C%20shape%20%3D%20%22invhouse%22%5D%3B%0A%20%20%20%20BB04%20%5Blabel%20%3D%20%22BB04%5Cn%5BIND%5D%20%3E%20V01%5Cn%5Cn%20%20%201.00%22%5D%3B%0A%20%20%20%20BB05%20%5Blabel%20%3D%20%22BB05%5CnV03%20%3D%3D%200%5Cn%5Cn%20%20%201.00%22%5D%3B%0A%20%20%20%20BB06%20%5Blabel%20%3D%20%22BB06%5Cn%5Cn%20%20%201.00%22%2C%20shape%20%3D%20%22invhouse%22%5D%3B%0A%20%20%20%20BB07%20%5Blabel%20%3D%20%22BB07%5Cn%5Cn%20%20%200.00%22%5D%3B%0A%20%20%20%20BB08%20%5Blabel%20%3D%20%22BB08%5Cn%5Cn%20%20%200.00%22%2C%20shape%20%3D%20%22note%22%5D%3B%0A%20%20%20%20BB09%20%5Blabel%20%3D%20%22BB09%5Cn%5Cn%20%20%200.00%22%2C%20shape%20%3D%20%22note%22%5D%3B%0A%20%20%20%20BB10%20%5Blabel%20%3D%20%22BB10%5Cn%5Cn%20%20%200.00%22%2C%20shape%20%3D%20%22note%22%5D%3B%0A%20%20%20%20BB11%20%5Blabel%20%3D%20%22BB11%5Cn%5BCAST%5D%20%3D%3D%200%5Cn%5Cn%20%20%200.00%22%5D%3B%0A%20%20%20%20BB12%20%5Blabel%20%3D%20%22BB12%5Cn%5Cn%20%20%200.00%22%5D%3B%0A%20%20%20%20BB13%20%5Blabel%20%3D%20%22BB13%5Cn%5Cn%20%20%200.00%22%5D%3B%0A%20%20%20%20BB14%20%5Blabel%20%3D%20%22BB14%5Cn%5Cn%20%20%200.00%22%2C%20shape%20%3D%20%22invhouse%22%5D%3B%0A%20%20%20%20BB15%20%5Blabel%20%3D%20%22BB15%5Cn%5BIND%5D%20%3E%20V01%5Cn%5Cn%20%20%200.00%22%5D%3B%0A%20%20%20%20BB16%20%5Blabel%20%3D%20%22BB16%5CnV06%20!%3D%200%5Cn%5Cn%20%20%200.00%22%5D%3B%0A%20%20%20%20BB17%20%5Blabel%20%3D%20%22BB17%5Cn%5Cn%20%20%200.00%22%5D%3B%0A%20%20%20%20BB18%20%5Blabel%20%3D%20%22BB18%5Cn%5Cn%20%20%200.00%22%2C%20shape%20%3D%20%22note%22%5D%3B%0A%20%20%20%20BB19%20%5Blabel%20%3D%20%22BB19%5Cn%5Cn%20%20%200.00%22%5D%3B%0A%20%20%20%20BB21%20%5Blabel%20%3D%20%22BB21%5Cn%5Cn%20%20%200.00%22%5D%3B%0A%20%20%20%20BB22%20%5Blabel%20%3D%20%22BB22%5Cn%5Cn%20%20%200.00%22%2C%20shape%20%3D%20%22note%22%5D%3B%0A%20%20%20%20BB20%20%5Blabel%20%3D%20%22BB20%5Cn%5Cn%20%20%200.00%22%2C%20shape%20%3D%20%22note%22%5D%3B%0A%20%20%20%20BB01%20-%3E%20BB02%20%5Bcolor%3Dblue%2C%20weight%3D20%2C%20label%3D%22%20%20%200.50%22%5D%3B%0A%20%20%20%20BB05%20-%3E%20BB02%20%5Bcolor%3Dgreen%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB08%20-%3E%20BB02%20%5Bcolor%3Dgreen%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB02%20-%3E%20BB03%20%5Bcolor%3Dblue%2C%20weight%3D20%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB02%20-%3E%20BB04%20%5Blabel%3D%22%20%20%201.00%22%5D%3B%0A%20%20%20%20BB04%20-%3E%20BB05%20%5Bcolor%3Dblue%2C%20weight%3D20%2C%20label%3D%22%20%20%201.00%22%5D%3B%0A%20%20%20%20BB05%20-%3E%20BB06%20%5Bcolor%3Dblue%2C%20weight%3D20%2C%20label%3D%22%20%20%201.00%22%5D%3B%0A%20%20%20%20BB04%20-%3E%20BB07%20%5Blabel%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB18%20-%3E%20BB08%20%5Bcolor%3Dgreen%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB20%20-%3E%20BB09%20%5Bcolor%3Dgreen%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB22%20-%3E%20BB10%20%5Bcolor%3Dgreen%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB17%20-%3E%20BB11%20%5Bcolor%3Dgreen%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB19%20-%3E%20BB11%20%5Bcolor%3Dgreen%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB21%20-%3E%20BB11%20%5Bcolor%3Dgreen%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB11%20-%3E%20BB12%20%5Bcolor%3Dblue%2C%20weight%3D20%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB11%20-%3E%20BB13%20%5Blabel%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB12%20-%3E%20BB13%20%5Bcolor%3Dblue%2C%20weight%3D20%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB09%20-%3E%20BB14%20%5Blabel%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB10%20-%3E%20BB14%20%5Blabel%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB07%20-%3E%20BB15%20%5Blabel%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB15%20-%3E%20BB16%20%5Bcolor%3Dblue%2C%20weight%3D20%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB16%20-%3E%20BB17%20%5Bcolor%3Dblue%2C%20weight%3D20%2C%20label%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB13%20-%3E%20BB18%20%5Blabel%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB16%20-%3E%20BB19%20%5Blabel%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB15%20-%3E%20BB21%20%5Blabel%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB13%20-%3E%20BB22%20%5Blabel%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB13%20-%3E%20BB20%20%5Blabel%3D%22%20%20%200.00%22%5D%3B%0A%20%20%20%20BB15%20-%3E%20BB11%20%5Bcolor%3Dred%5D%0A%20%20%20%20BB16%20-%3E%20BB11%20%5Bcolor%3Dred%5D%0A%20%20%20%20BB17%20-%3E%20BB11%20%5Bcolor%3Dred%5D%0A%20%20%20%20BB19%20-%3E%20BB11%20%5Bcolor%3Dred%5D%0A%20%20%20%20BB21%20-%3E%20BB11%20%5Bcolor%3Dred%5D%0A%20%20%20%20BB01%20-%3E%20BB02%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB02%20-%3E%20BB03%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB03%20-%3E%20BB04%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB04%20-%3E%20BB05%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB05%20-%3E%20BB06%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB06%20-%3E%20BB07%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB07%20-%3E%20BB08%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB08%20-%3E%20BB09%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB09%20-%3E%20BB10%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB10%20-%3E%20BB11%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB11%20-%3E%20BB12%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB12%20-%3E%20BB13%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB13%20-%3E%20BB14%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB14%20-%3E%20BB15%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB15%20-%3E%20BB16%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB16%20-%3E%20BB17%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB17%20-%3E%20BB18%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB18%20-%3E%20BB19%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB19%20-%3E%20BB21%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB21%20-%3E%20BB22%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20BB22%20-%3E%20BB20%20%5Bstyle%3D%22invis%22%2C%20weight%3D25%5D%3B%0A%20%20%20%20subgraph%20cluster_0%20%7B%0A%20%20%20%20%20%20%20%20label%20%3D%20%22L00%22%3B%0A%20%20%20%20%20%20%20%20color%20%3D%20blue%3B%0A%20%20%20%20%20%20%20%20BB02%3BBB04%3BBB07%3BBB15%3BBB21%3BBB16%3BBB19%3BBB17%3BBB11%3BBB12%3BBB13%3BBB18%3BBB08%3BBB05%3B%0A%20%20%20%20%7D%0A%7D%0A); red edges are EH successor edges. The `CALLFINALLY` is `BB19` and the `CALLFINALLYRET` is `BB20`.